src/meta/meta_turtle.c

```diff
@@ -80,129 +92,180 @@ __metadata_backup_target_uri_search(
+                meta_state->max_len = key->size;
+
+            WT_RET(__wt_realloc_def(session, &meta_state->allocated, meta_state->slot + 2,
+              &meta_state->partial_backup_names));
+            p = &meta_state->partial_backup_names[meta_state->slot];
+            p[0] = p[1] = NULL;
+
+            WT_RET(__wt_strndup(session, (char *)key->data, key->size,
+              &meta_state->partial_backup_names[meta_state->slot]));
+            meta_state->slot++;
+        }
     }
 
+    /*
+     * In the case of partial backup restore, add the entry to the metadata even if the table entry
+     * doesn't exist so that we can correctly drop all related entries via the schema code later.
+     */
+    WT_RET(__wt_metadata_update(session, key->data, value->data));
+
+    return (0);
+}
+
+/*
+ * __metadata_load_hot_backup --
+ *     Load the contents of any hot backup file.
+ */
+static int
+__metadata_load_hot_backup(WT_SESSION_IMPL *session, WT_BACKUPHASH *backuphash)
+{
+    WT_CONFIG_ITEM cval;
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+    WT_METADATA_FILE_WALK_STATE meta_state;
+    size_t file_len;
+    char *filename, *metadata_conf, *tablename;
+    const char *drop_cfg[] = {WT_CONFIG_BASE(session, WT_SESSION_drop), "remove_files=false", NULL};
+    bool exist;
+
+    file_len = 0;
+    conn = S2C(session);
+    filename = metadata_conf = tablename = NULL;
+    exist = false;
+
+    WT_CLEAR(meta_state);
+    meta_state.backuphash = backuphash;
+
+    /* Open the metadata backup file and iterate over the key value pairs. */
+    WT_ERR(__wt_read_metadata_file(
+      session, WT_METADATA_BACKUP, __metadata_entry_worker, &meta_state, &exist));
+    if (!exist)
+        goto err;
+
     F_SET(conn, WT_CONN_WAS_BACKUP);
-    if (F_ISSET(conn, WT_CONN_BACKUP_PARTIAL_RESTORE) && partial_backup_names != NULL) {
-        WT_ERR(__wt_calloc_def(session, slot + 1, &conn->partial_backup_remove_ids));
-        file_len = strlen("file:") + max_len + strlen(".wt") + 1;
+    if (F_ISSET(conn, WT_CONN_BACKUP_PARTIAL_RESTORE) && meta_state.partial_backup_names != NULL) {
+        WT_ERR(__wt_calloc_def(session, meta_state.slot + 1, &conn->partial_backup_remove_ids));
+        file_len = strlen("file:") + meta_state.max_len + strlen(".wt") + 1;
         WT_ERR(__wt_calloc_def(session, file_len, &filename));
         /*
          * Parse through the partial backup list and attempt to clean up all metadata references
          * relating to the file. To do so, perform a schema drop operation on the table to cleanly
          * remove all linked references. At the same time generate a list of btree ids to be used in
          * recovery to truncate all the history store records.
          */
-        for (slot = 0; partial_backup_names[slot] != NULL; ++slot) {
-            tablename = partial_backup_names[slot];
+        for (meta_state.slot = 0; meta_state.partial_backup_names[meta_state.slot] != NULL;
```

Comment:
np: Reading back the original code, it feels uneasy re-using the slot variable, incase we accidentally misuse the slot value. could we create another variable **i** instead here that traverses the loop?

Reply:
Done. I've also replaced condition check in the loop from `meta_state.partial_backup_names[i] != NULL` to `i < meta_state.slot`.